internal/v2/tls_config_store/tls_config_store.go

```diff
@@ -107,8 +107,89 @@ func GetTlsConfigurationForClient(serverHostname string, cstream s2av2pb.S2AServ
 
 // GetTlsConfigurationForServer returns a tls.Config instance for use by a server application.
 func GetTlsConfigurationForServer(cstream s2av2pb.S2AService_SetUpSessionClient, tokenManager tokenmanager.AccessTokenManager, localIdentities []*commonpbv1.Identity) (*tls.Config, error) {
-	// TODO(rmehta19): Add an extra auth mechanism for the SNI.
-	// TODO(rmehta19): move call to S2Av2 to a helper function.
+	return &tls.Config {
+		GetConfigForClient: ClientConfig(tokenManager, localIdentities, cstream),
+	}, nil
+}
+
+// ClientConfig builds a TLS config for a server to establish a secure
+// connection with a client, based on SNI communicated during ClientHello.
+// Ensures that server presents the correct certificate to establish a TLS
+// connection.
+func ClientConfig(tokenManager tokenmanager.AccessTokenManager, localIdentities []*commonpbv1.Identity, cstream s2av2pb.S2AService_SetUpSessionClient) func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+	return func(chi *tls.ClientHelloInfo) (*tls.Config, error) {
+		localIdentities = append(localIdentities,
+		&commonpbv1.Identity{
+			IdentityOneof: &commonpbv1.Identity_Hostname{
+				Hostname: chi.ServerName,
+			},
+		})
+		tlsConfig, err := getServerConfigFromS2Av2(tokenManager, localIdentities, cstream)
+		if err != nil {
+			return nil, err
+		}
+
+		var cert tls.Certificate
+		for i, v := range tlsConfig.CertificateChain {
+			// Populate Certificates field
+			block, _ := pem.Decode([]byte(v))
+			if block == nil {
+				return nil, errors.New("certificate in CertificateChain obtained from S2Av2 is empty.")
+			}
+			x509Cert, err := x509.ParseCertificate(block.Bytes)
+			if err != nil {
+				return nil, err
+			}
+			cert.Certificate = append(cert.Certificate, x509Cert.Raw)
+			if i == 0 {
+				cert.Leaf = x509Cert
+			}
+		}
+
+		cert.PrivateKey = remotesigner.New(cert.Leaf, cstream, &commonpbv1.Identity {
```

Comment:
Now that we only send auth mechanisms in the GetTlsConfigurationRequest, should we remove the local identity from the remote signer? IIUC, it is no longer needed.

Reply:
In remote signer, I use the localIdentity field of the SessionReq to determine who is requesting the signing operation, a client using s2a or a server using s2a. This field is used on the fakes2av2 side to determine which private key to use to perform the signing. 

Perhaps in the next PR, when I add the token managing feature to fakes2av2, I can use that field to determine who is using s2av2.